k9mail/src/main/java/com/fsck/k9/activity/compose/AttachmentPresenter.java

```diff
@@ -0,0 +1,371 @@
+                initAttachmentContentLoader(attachment);
+            }
+        }
+    }
+
+    public boolean checkOkForSendingOrDraftSaving() {
+        if (actionToPerformAfterWaiting != WaitingAction.NONE) {
+            return true;
+        }
+
+        if (hasLoadingAttachments()) {
+            actionToPerformAfterWaiting = WaitingAction.SEND;
+            attachmentMvpView.showWaitingForAttachmentDialog(actionToPerformAfterWaiting);
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean hasLoadingAttachments() {
+        for (Attachment attachment : attachments.values()) {
+            Loader loader = loaderManager.getLoader(attachment.loaderId);
+            if (loader != null && loader.isStarted()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public ArrayList<Attachment> createAttachmentList() {
+        ArrayList<Attachment> result = new ArrayList<>();
+        for (Attachment attachment : attachments.values()) {
+            result.add(attachment);
+        }
+        return result;
+    }
+
+    public void onClickAddAttachment(RecipientPresenter recipientPresenter) {
+        AttachErrorState maybeAttachErrorState = recipientPresenter.getCurrentCryptoStatus().getAttachErrorStateOrNull();
+        if (maybeAttachErrorState != null) {
+            recipientPresenter.showPgpAttachError(maybeAttachErrorState);
+            return;
+        }
+
+        attachmentMvpView.showPickAttachmentDialog(REQUEST_CODE_ATTACHMENT_URI);
+    }
+
+    private void addAttachment(Uri uri) {
+        addAttachment(uri, null);
+    }
+
+    public void addAttachment(Uri uri, String contentType) {
+        Attachment attachment = new Attachment();
+        attachment.uri = uri;
+        attachment.state = Attachment.LoadingState.URI_ONLY;
+        attachment.contentType = contentType;
+        attachment.loaderId = getNextFreeLoaderId();
+
+        if (attachments.containsKey(uri)) {
+            return;
+        }
+
+        attachments.put(uri, attachment);
+        attachmentMvpView.addAttachmentView(attachment);
+
+        initAttachmentInfoLoader(attachment);
+    }
+
+    private void initAttachmentInfoLoader(Attachment attachment) {
+        Bundle bundle = new Bundle();
+        bundle.putParcelable(LOADER_ARG_ATTACHMENT, attachment.uri);
+
+        loaderManager.initLoader(attachment.loaderId, bundle, mAttachmentInfoLoaderCallback);
+    }
+
+    private void initAttachmentContentLoader(Attachment attachment) {
+        Bundle bundle = new Bundle();
+        bundle.putParcelable(LOADER_ARG_ATTACHMENT, attachment.uri);
+
+        loaderManager.initLoader(attachment.loaderId, bundle, mAttachmentContentLoaderCallback);
+    }
+
+    private int getNextFreeLoaderId() {
+        if (nextLoaderId >= MAX_TOTAL_LOADERS) {
+            throw new AssertionError("more than " + MAX_TOTAL_LOADERS + " attachments? hum.");
```

Comment:
Might want to move to a less radical solution here? Not sure. Does "more than 128 attachments" ever happen? Should we display a toast or something instead?